File: play-ahc-ws-standalone/src/main/java/play/libs/oauth/OAuth.java
/*
 * Copyright (C) 2009-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.oauth;

import play.shaded.oauth.oauth.signpost.OAuthConsumer;
import play.shaded.oauth.oauth.signpost.OAuthProvider;
import play.shaded.oauth.oauth.signpost.basic.DefaultOAuthConsumer;
import play.shaded.oauth.oauth.signpost.basic.DefaultOAuthProvider;
import play.shaded.oauth.oauth.signpost.exception.OAuthException;
import play.shaded.ahc.org.asynchttpclient.oauth.OAuthSignatureCalculator;
import play.libs.ws.WSSignatureCalculator;

public class OAuth {

    private ServiceInfo info;
    private OAuthProvider provider;

    public OAuth(ServiceInfo info) {
        this(info, true);
    }

    public OAuth(ServiceInfo info, boolean use10a) {
        this.info = info;
        this.provider = new DefaultOAuthProvider(info.requestTokenURL, info.accessTokenURL, info.authorizationURL);
        this.provider.setOAuth10a(use10a);
    }

    public ServiceInfo getInfo() {
        return info;
    }

    public OAuthProvider getProvider() {
        return provider;
    }

    /**
     * Request the request token and secret.
     *
     * @param callbackURL the URL where the provider should redirect to (usually a URL on the current app)
     * @return A Right(RequestToken) in case of success, Left(OAuthException) otherwise
     */
    public RequestToken retrieveRequestToken(String callbackURL) {
        OAuthConsumer consumer = new DefaultOAuthConsumer(info.key.key, info.key.secret);
        try {
            provider.retrieveRequestToken(consumer, callbackURL);
            return new RequestToken(consumer.getToken(), consumer.getTokenSecret());
        } catch (OAuthException ex) {
            throw new RuntimeException(ex);
        }
    }

    /**
     * Exchange a request token for an access token.
     *
     * @param token    the token/secret pair obtained from a previous call
     * @param verifier a string you got through your user, with redirection
     * @return A Right(RequestToken) in case of success, Left(OAuthException) otherwise
     */
    public RequestToken retrieveAccessToken(RequestToken token, String verifier) {
        OAuthConsumer consumer = new DefaultOAuthConsumer(info.key.key, info.key.secret);
        consumer.setTokenWithSecret(token.token, token.secret);
        try {
            provider.retrieveAccessToken(consumer, verifier);
            return new RequestToken(consumer.getToken(), consumer.getTokenSecret());
        } catch (OAuthException ex) {
            throw new RuntimeException(ex);
        }
    }

    /**
     * The URL where the user needs to be redirected to grant authorization to your application.
     *
     * @param token request token
     * @return the url
     */
    public String redirectUrl(String token) {
        return play.shaded.oauth.oauth.signpost.OAuth.addQueryParameters(
                provider.getAuthorizationWebsiteUrl(),
                play.shaded.oauth.oauth.signpost.OAuth.OAUTH_TOKEN,
                token
        );
    }

    /**
     * A consumer key / consumer secret pair that the OAuth provider gave you, to identify your application.
     */
    public static class ConsumerKey {
        public String key;
        public String secret;

        public ConsumerKey(String key, String secret) {
            this.key = key;
            this.secret = secret;
        }
    }

    /**
     * A request token / token secret pair, to be used for a specific user.
     */
    public static class RequestToken {
        public String token;
        public String secret;

        public RequestToken(String token, String secret) {
            this.token = token;
            this.secret = secret;
        }
    }

    /**
     * The information identifying a oauth provider: URLs and the consumer key / consumer secret pair.
     */
    public static class ServiceInfo {
        public String requestTokenURL;
        public String accessTokenURL;
        public String authorizationURL;
        public ConsumerKey key;

        public ServiceInfo(String requestTokenURL, String accessTokenURL, String authorizationURL, ConsumerKey key) {
            this.requestTokenURL = requestTokenURL;
            this.accessTokenURL = accessTokenURL;
            this.authorizationURL = authorizationURL;
            this.key = key;
        }
    }

    /**
     * A signature calculator for the Play WS API.
     * <p>
     * Example:
     * {{{
     * WS.url("http://example.com/protected").sign(OAuthCalculator(service, token)).get()
     * }}}
     */
    public static class OAuthCalculator implements WSSignatureCalculator {

        private OAuthSignatureCalculator calculator;

        public OAuthCalculator(ConsumerKey consumerKey, RequestToken token) {
            play.shaded.ahc.org.asynchttpclient.oauth.ConsumerKey ahcConsumerKey = new play.shaded.ahc.org.asynchttpclient.oauth.ConsumerKey(consumerKey.key, consumerKey.secret);
            play.shaded.ahc.org.asynchttpclient.oauth.RequestToken ahcRequestToken = new play.shaded.ahc.org.asynchttpclient.oauth.RequestToken(token.token, token.secret);
            calculator = new OAuthSignatureCalculator(ahcConsumerKey, ahcRequestToken);
        }

        public OAuthSignatureCalculator getCalculator() {
            return calculator;
        }
    }

}

File: play-ahc-ws-standalone/src/main/scala/play/api/libs/oauth/OAuth.scala
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.oauth

import play.shaded.oauth.oauth.signpost.basic.DefaultOAuthConsumer
import play.shaded.oauth.oauth.signpost.basic.DefaultOAuthProvider
import play.shaded.oauth.oauth.signpost.exception.OAuthException
import play.shaded.ahc.org.asynchttpclient.oauth.OAuthSignatureCalculator
import play.shaded.ahc.org.asynchttpclient.Request
import play.shaded.ahc.org.asynchttpclient.RequestBuilderBase
import play.shaded.ahc.org.asynchttpclient.SignatureCalculator
import play.api.libs.ws.WSSignatureCalculator

import play.shaded.ahc.org.asynchttpclient.oauth.{ ConsumerKey => AHCConsumerKey }
import play.shaded.ahc.org.asynchttpclient.oauth.{ RequestToken => AHCRequestToken }

/**
 * Library to access resources protected by OAuth 1.0a.
 *
 * @param info   the service information, including the required URLs and the application id and secret
 * @param use10a whether the service should use the 1.0 version of the spec, or the 1.0a version fixing a security issue.
 *               You must use the version corresponding to the
 */
case class OAuth(info: ServiceInfo, use10a: Boolean = true) {

  private val provider = {
    val p = new DefaultOAuthProvider(info.requestTokenURL, info.accessTokenURL, info.authorizationURL)
    p.setOAuth10a(use10a)
    p
  }

  /**
   * Request the request token and secret.
   *
   * @param callbackURL the URL where the provider should redirect to (usually a URL on the current app)
   * @return A Right(RequestToken) in case of success, Left(OAuthException) otherwise
   */
  def retrieveRequestToken(callbackURL: String): Either[OAuthException, RequestToken] = {
    val consumer = new DefaultOAuthConsumer(info.key.key, info.key.secret)
    try {
      provider.retrieveRequestToken(consumer, callbackURL)
      Right(RequestToken(consumer.getToken(), consumer.getTokenSecret()))
    } catch {
      case e: OAuthException => Left(e)
    }
  }

  /**
   * Exchange a request token for an access token.
   *
   * @param token    the token/secret pair obtained from a previous call
   * @param verifier a string you got through your user, with redirection
   * @return A Right(RequestToken) in case of success, Left(OAuthException) otherwise
   */
  def retrieveAccessToken(token: RequestToken, verifier: String): Either[OAuthException, RequestToken] = {
    val consumer = new DefaultOAuthConsumer(info.key.key, info.key.secret)
    consumer.setTokenWithSecret(token.token, token.secret)
    try {
      provider.retrieveAccessToken(consumer, verifier)
      Right(RequestToken(consumer.getToken(), consumer.getTokenSecret()))
    } catch {
      case e: OAuthException => Left(e)
    }
  }

  /**
   * The URL where the user needs to be redirected to grant authorization to your application.
   *
   * @param token request token
   */
  def redirectUrl(token: String): String = {
    import play.shaded.oauth.oauth.signpost.{ OAuth => O }
    O.addQueryParameters(
      provider.getAuthorizationWebsiteUrl(),
      O.OAUTH_TOKEN,
      token
    )
  }

}

/**
 * A consumer key / consumer secret pair that the OAuth provider gave you, to identify your application.
 */
case class ConsumerKey(key: String, secret: String)

/**
 * A request token / token secret pair, to be used for a specific user.
 */
case class RequestToken(token: String, secret: String)

/**
 * The information identifying a oauth provider: URLs and the consumer key / consumer secret pair.
 */
case class ServiceInfo(requestTokenURL: String, accessTokenURL: String, authorizationURL: String, key: ConsumerKey)

/**
 * The public AsyncHttpClient implementation of WSSignatureCalculator.
 */
class OAuthCalculator(consumerKey: ConsumerKey, requestToken: RequestToken)
    extends WSSignatureCalculator
    with SignatureCalculator {

  private val ahcConsumerKey  = new AHCConsumerKey(consumerKey.key, consumerKey.secret)
  private val ahcRequestToken = new AHCRequestToken(requestToken.token, requestToken.secret)
  private val calculator      = new OAuthSignatureCalculator(ahcConsumerKey, ahcRequestToken)

  override def calculateAndAddSignature(request: Request, requestBuilder: RequestBuilderBase[_]): Unit = {
    calculator.calculateAndAddSignature(request, requestBuilder)
  }
}

/**
 * Object for creating signature calculator for the Play WS API.
 *
 * Example:
 * {{{
 * import play.api.libs.oauth.{ ConsumerKey, OAuthCalculator, RequestToken }
 * import play.api.libs.ws.ahc.StandaloneAhcWSClient
 *
 * def example(
 *   twitterConsumerKey: String,
 *   twitterConsumerSecret: String,
 *   accessTokenKey: String,
 *   accessTokenSecret: String,
 *   ws: StandaloneAhcWSClient) = {
 *   val consumerKey: ConsumerKey =
 *     ConsumerKey(twitterConsumerKey, twitterConsumerSecret)
 *
 *   val requestToken: RequestToken =
 *     RequestToken(accessTokenKey, accessTokenSecret)
 *
 *   ws.url("http://example.com/protected").
 *     sign(OAuthCalculator(consumerKey, requestToken)).get()
 * }
 * }}}
 */
object OAuthCalculator {
  def apply(consumerKey: ConsumerKey, token: RequestToken): WSSignatureCalculator = {
    new OAuthCalculator(consumerKey, token)
  }
}
